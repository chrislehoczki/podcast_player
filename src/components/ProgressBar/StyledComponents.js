import styled from "styled-components";
import theme from "../../styles/theme";

const sliderThumbStyles = props => `
  width: 15px;
  height: 15px;
  border-radius: 100%;
  background: ${theme.primary};
`;

export const Container = styled.div`
  position: absolute;
  bottom: 25px;
  width: 80%;
  margin: 0;
  display: flex;
  align-items: center;
  color: ${theme.primary};
  .slider {
    flex: 6;
    -webkit-appearance: none;
    width: 100%;
    height: 5px;
    border-radius: 5px;
    background: ${theme.primary};
    outline: none;
    &::-webkit-slider-thumb {
      -webkit-appearance: none;
      appearance: none;
      ${props => sliderThumbStyles(props)}
    }
    &::-moz-range-thumb {
      ${props => sliderThumbStyles(props)}
    }
  }
`;
