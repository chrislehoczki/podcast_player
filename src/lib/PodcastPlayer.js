function PodcastPlayer() {}

PodcastPlayer.prototype = {
  audio: null,
  progressInterval: null,
  play() {
    this.audio.play();
  },
  pause() {
    this.audio.pause();
  },
  stop() {
    if (this.audio) {
      this.pause();
      this.audio.currentTime = 0;
    }
  },
  load(url, loadCb, endCb) {
    this.stop();
    this.loading = true;
    this.audio = new window.Audio(url);
    this.audio.preload = "auto";
    this.audio.onended = endCb;
    this.audio.addEventListener("canplaythrough", loadCb);
  },
  seek(val) {
    this.audio.currentTime = val;
  },
  createProgressInterval(cb) {
    if (!this.progressInterval) {
      this.progressInterval = setInterval(() => {
        if (this.audio) {
          const duration = this.audio.duration > 0 ? this.audio.duration : 1000;
          const progress = +(this.audio.currentTime / duration).toFixed(3);
          cb(progress);
        }
      }, 100);
    }
  },
  destroyProgressInterval() {
    clearInterval(this.progressInterval);
    this.progressInterval = null;
  }
};

const podcastPlayer = new PodcastPlayer();
export default podcastPlayer;
