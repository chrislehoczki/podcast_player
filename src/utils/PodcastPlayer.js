function PodcastPlayer() {}

PodcastPlayer.prototype = {
  audio: null,
  progressInterval: null,
  play() {
    this.audio.play();
  },
  pause() {
    this.audio.pause();
  },
  stop() {
    if (this.audio) {
      this.pause();
      this.audio.currentTime = 0;
    }
  },
  load(url, loadCb, endCb) {
    this.stop();
    this.loading = true;
    this.audio = new window.Audio(url);
    this.audio.preload = "auto";
    this.audio.onended = endCb;
    // this.audio.currentTime = 10000;
    this.audio.addEventListener("canplaythrough", loadCb);
  },
  createProgressInterval(cb) {
    if (!this.progressInterval) {
      this.progressInterval = setInterval(() => {
        if (this.audio) {
          const progress = this.audio.currentTime / this.audio.duration;
          cb(progress);
        }
      }, 100);
    }
  },
  destroyProgressInterval() {
    clearInterval(this.progressInterval);
    this.progressInterval = null;
  }
};

const podcastPlayer = new PodcastPlayer();
export default podcastPlayer;
