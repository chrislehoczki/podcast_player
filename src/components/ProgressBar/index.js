import React, { useEffect, useState } from "react";
import podcastPlayer from "../../lib/PodcastPlayer";
import { Container } from "./StyledComponents";

const ProgressBar = () => {
  const [progress, setProgress] = useState(0);
  useEffect(() => {
    podcastPlayer.createProgressInterval(progress => {
      setProgress(+progress);
    });
  }, [setProgress]);
  return (
    <Container>
      <input
        type="range"
        min={0}
        max={1}
        value={progress || 0}
        step={0.001}
        className="slider"
        readOnly={true}
      />
    </Container>
  );
};

export default ProgressBar;
