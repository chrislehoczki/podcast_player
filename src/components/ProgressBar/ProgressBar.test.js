import React from "react";
import { shallow, mount } from "enzyme";
import ProgressBar from "./index";
import PodcastPlayer from "../../lib/PodcastPlayer";
import mockData from "../../test/mock_data";

describe("<Loader />", () => {
  it("Renders correctly", () => {
    const component = shallow(<ProgressBar />);
    expect(component).toMatchSnapshot();
  });
  it("Renders a an input slider", () => {
    const component = shallow(<ProgressBar />);
    expect(component.find("input")).toExist();
  });
  it("Slider shows the correct value of the state", done => {
    let component = mount(<ProgressBar size={60} />);
    const input = component.find("input");
    expect(input.instance().value).toEqual("0");
    PodcastPlayer.load(mockData[0].url, () => {});
    PodcastPlayer.seek(562);
    setTimeout(() => {
      expect(input.instance().value).toEqual("0.562");
      done();
    }, 100);
  });
});
