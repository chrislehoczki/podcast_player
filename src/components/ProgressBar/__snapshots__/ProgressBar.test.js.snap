// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Loader /> Renders correctly 1`] = `
<styled.div>
  <input
    className="slider"
    max={1}
    min={0}
    readOnly={true}
    step={0.001}
    type="range"
    value={0}
  />
</styled.div>
`;
